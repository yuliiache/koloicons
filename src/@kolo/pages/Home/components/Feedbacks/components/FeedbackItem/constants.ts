export const FEEDBACK_TEXT_TEST_ID = 'feedbackText';
