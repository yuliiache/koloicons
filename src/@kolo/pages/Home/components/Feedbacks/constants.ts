export const FEEDBACKS_KEY = 'feedbacks';
