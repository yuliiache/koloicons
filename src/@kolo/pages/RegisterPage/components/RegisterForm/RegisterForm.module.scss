@import "@kolo/design";

.registerForm {
  margin-bottom: 24px;
}

.checkbox {
  &Wrapper {
    display: flex;
    justify-content: flex-start;

    &:not(:last-of-type) {
      margin-bottom: 8px;
    }

    &:last-of-type {
      margin-bottom: 16px;
    }
  }

  &Text {
    @include main-text(0.7rem, 1.43, null);
    color: $secFontColor;

    &.textDesktop {
      display: block;

      @media (max-width: $mobileWidth) {
        display: none;
      }
    }

    &.textMobile {
      display: none;

      @media (max-width: $mobileWidth) {
        display: block;
      }
    }

    & > a {
      @include main-text(0.7rem, 1.43, null);
      color: $secFontColor;
    }
  }
}
