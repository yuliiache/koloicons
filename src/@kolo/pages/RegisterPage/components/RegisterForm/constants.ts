export const REGISTER_FORM_TEST_ID = 'registerFormComponent';

export const enum CheckboxDataTestId {
  SUBSCRIBE_AGREEMENT_CHECKBOX = 'subscribeAgreementCheckbox',
  TERMS_AGREEMENT_CHECKBOX = 'termsAgreementCheckbox',
}
