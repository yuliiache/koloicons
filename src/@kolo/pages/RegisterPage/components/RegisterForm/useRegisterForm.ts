import {FormApi} from 'final-form';

import {RegisterFormValues} from './types/RegisterFormValues';

interface RegisterFormHookReturn {
  initialValues: RegisterFormValues;
  onFormSubmit: (values: RegisterFormValues, form: FormApi<RegisterFormValues, RegisterFormValues>) => void;
}

const initialValues: RegisterFormValues = {
  firstName: '',
  lastName: '',
  email: '',
  password: '',
  subscribeAgreement: true,
  termsAgreement: false,
};

export const useRegisterForm = (onSubmit: (values: RegisterFormValues) => void): RegisterFormHookReturn => {
  const onFormSubmit = (values: RegisterFormValues, form: FormApi<RegisterFormValues, RegisterFormValues>): void => {
    onSubmit(values);
    form.reset();

    /* reset field's flags (e.g. modified, touched, etc.) to their initial state
    to display errors correctly */
    form.batch(() => {
      Object.keys(initialValues).forEach((fieldName: string) => {
        form.resetFieldState(fieldName as keyof RegisterFormValues);
      });
    });
  };

  return {initialValues, onFormSubmit};
};
