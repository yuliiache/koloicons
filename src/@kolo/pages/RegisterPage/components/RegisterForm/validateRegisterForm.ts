import {
  validateCheckbox,
  validateEmail,
  validateFirstName,
  validateLastName,
  validatePassword,
} from '@kolo/services/validators/commonValidators';

import {RegisterFormValues} from './types/RegisterFormValues';

interface ValidationErrors {
  firstName?: string;
  lastName?: string;
  email?: string;
  password?: string;
  termsAgreement?: boolean;
}

export const validateRegisterForm = (values: RegisterFormValues): ValidationErrors => {
  const errors: ValidationErrors = {};
  const {firstName, lastName, email, password, termsAgreement} = values;

  errors.firstName = validateFirstName(firstName);
  errors.lastName = validateLastName(lastName);
  errors.email = validateEmail(email);
  errors.password = validatePassword(password);
  errors.termsAgreement = validateCheckbox(termsAgreement);

  return errors;
};
