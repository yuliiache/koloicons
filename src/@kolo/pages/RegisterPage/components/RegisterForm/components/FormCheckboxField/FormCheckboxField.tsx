import {Checkbox} from '@kolo/uiKit/Checkbox/Checkbox';
import {FC} from 'react';
import {Field, FieldRenderProps} from 'react-final-form';

import {CheckboxDataTestId} from '../../constants';

interface Props {
  dataTestId: CheckboxDataTestId;
  fieldName: string;
}

const FormCheckboxField: FC<Props> = ({dataTestId, fieldName}) => {
  return (
    <div data-testid={dataTestId}>
      <Field
        name={fieldName}
        type="checkbox"
      >
        {({input: {checked, onChange}, meta: {modified, error, touched}}: FieldRenderProps<boolean>) => (
          <Checkbox
            id={fieldName}
            onChange={onChange}
            isChecked={checked || false}
            hasError={(modified && !checked && error) || (touched && error)}
          />
        )}
      </Field>
    </div>
  );
};

export default FormCheckboxField;
