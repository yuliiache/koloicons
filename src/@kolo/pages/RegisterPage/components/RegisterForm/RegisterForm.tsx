import Button from '@kolo/uiKit/Button/Button';
import {ButtonSize, ButtonType} from '@kolo/uiKit/Button/constants';
import {InputType} from '@kolo/uiKit/Input/InputType';
import Link from '@kolo/uiKit/Link/Link';
import classNames from 'classnames';
import {AppRoute} from 'constants/AppRoute';
import {FC} from 'react';
import {Form, FormRenderProps} from 'react-final-form';

import FormCheckboxField from './components/FormCheckboxField/FormCheckboxField';
import FormInputField from './components/FormInputField/FormInputField';
import {CheckboxDataTestId, REGISTER_FORM_TEST_ID} from './constants';
import styles from './RegisterForm.module.scss';
import {RegisterFormValues} from './types/RegisterFormValues';
import {useRegisterForm} from './useRegisterForm';
import {validateRegisterForm} from './validateRegisterForm';

interface Props {
  onSubmit: (values: RegisterFormValues) => void;
}

const RegisterForm: FC<Props> = ({onSubmit}) => {
  const {initialValues, onFormSubmit} = useRegisterForm(onSubmit);

  return (
    <div data-testid={REGISTER_FORM_TEST_ID}>
      <Form
        onSubmit={onFormSubmit}
        initialValues={initialValues}
        subscription={{submitError: true}}
        validate={validateRegisterForm}
        render={(props: FormRenderProps<RegisterFormValues>) => (
          <form
            className={styles.registerForm}
            onSubmit={props.handleSubmit}
          >
            <FormInputField
              fieldName={'firstName'}
              label={'First name'}
              type={InputType.TEXT}
            />
            <FormInputField
              fieldName={'lastName'}
              label={'Last name'}
              type={InputType.TEXT}
            />
            <FormInputField
              fieldName={'email'}
              label={'Email'}
              type={InputType.TEXT}
            />
            <FormInputField
              fieldName={'password'}
              label={'Password'}
              type={InputType.PASSWORD}
            />

            <div className={styles.checkboxWrapper}>
              <FormCheckboxField
                dataTestId={CheckboxDataTestId.SUBSCRIBE_AGREEMENT_CHECKBOX}
                fieldName="subscribeAgreement"
              />

              <p className={styles.checkboxText}>
                Agree to receive emails from koloicons with discounts or special offers.
              </p>
            </div>

            <div className={styles.checkboxWrapper}>
              <FormCheckboxField
                dataTestId={CheckboxDataTestId.TERMS_AGREEMENT_CHECKBOX}
                fieldName="termsAgreement"
              />

              <p className={classNames(styles.checkboxText, styles.textDesktop)}>
                Agree to KOLO Icons Terms of Use and Privacy Policy.
              </p>
              <p className={classNames(styles.checkboxText, styles.textMobile)}>
                Agree to KOLO Icons{' '}
                <Link url={AppRoute.TERMS_OF_SERVICE}>
                  <span>Terms of Use</span>
                </Link>{' '}
                and{' '}
                <Link url={AppRoute.PRIVACY_POLICY}>
                  <span>Privacy Policy</span>
                </Link>
                .
              </p>
            </div>

            <Button
              size={ButtonSize.FULL_WIDTH}
              text="Sign Up"
              type={ButtonType.SUBMIT}
            />
          </form>
        )}
      />
    </div>
  );
};

export default RegisterForm;
