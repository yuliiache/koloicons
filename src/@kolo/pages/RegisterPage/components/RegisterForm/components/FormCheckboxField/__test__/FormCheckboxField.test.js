import {Form} from 'react-final-form';
import {CheckboxDataTestId} from '@kolo/pages/RegisterPage/components/RegisterForm/constants';
import {render, screen} from '@testing-library/react';

import FormCheckboxField from '../FormCheckboxField';

describe('FormCheckboxField component', () => {
  const onSubmitMock = jest.fn();
  const formProps = {
    initialValues: {
      subscribeAgreement: true,
      termsAgreement: false,
    },
    onSubmit: onSubmitMock,
  };

  it('should render snapshot successfully for checked subscription agreement checkbox', () => {
    const dataTestId = CheckboxDataTestId.SUBSCRIBE_AGREEMENT_CHECKBOX;
    const fieldProps = {
      dataTestId: dataTestId,
      fieldName: 'subscribeAgreement',
    };
    const {asFragment} = render(<Form {...formProps}>{() => <FormCheckboxField {...fieldProps} />}</Form>);

    const checkboxWrapper = screen.getByTestId(dataTestId);
    const subscribeAgreementCheckbox = checkboxWrapper.querySelector('input[type="checkbox"]');

    expect(subscribeAgreementCheckbox).toBeChecked();

    expect(asFragment()).toMatchSnapshot();
  });

  it('should render snapshot successfully for unchecked terms agreement checkbox', () => {
    const dataTestId = CheckboxDataTestId.TERMS_AGREEMENT_CHECKBOX;
    const fieldProps = {
      dataTestId: dataTestId,
      fieldName: 'termsAgreement',
    };
    const {asFragment} = render(<Form {...formProps}>{() => <FormCheckboxField {...fieldProps} />}</Form>);

    const checkboxWrapper = screen.getByTestId(dataTestId);
    const termsAgreementCheckbox = checkboxWrapper.querySelector('input[type="checkbox"]');

    expect(termsAgreementCheckbox).not.toBeChecked();

    expect(asFragment()).toMatchSnapshot();
  });
});
