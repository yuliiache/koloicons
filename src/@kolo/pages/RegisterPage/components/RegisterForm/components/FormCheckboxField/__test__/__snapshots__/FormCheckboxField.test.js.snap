// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FormCheckboxField component should render snapshot successfully for checked subscription agreement checkbox 1`] = `
<DocumentFragment>
  <div
    data-testid="subscribeAgreementCheckbox"
  >
    <label
      class="checkbox"
    >
      <input
        checked=""
        class="checkbox__input"
        id="subscribeAgreement"
        type="checkbox"
      />
      <span
        class="checkbox__indicator"
      />
    </label>
  </div>
</DocumentFragment>
`;

exports[`FormCheckboxField component should render snapshot successfully for unchecked terms agreement checkbox 1`] = `
<DocumentFragment>
  <div
    data-testid="termsAgreementCheckbox"
  >
    <label
      class="checkbox"
    >
      <input
        class="checkbox__input"
        id="termsAgreement"
        type="checkbox"
      />
      <span
        class="checkbox__indicator"
      />
    </label>
  </div>
</DocumentFragment>
`;
